Mainnet/STOContract/STOContract/STOContract.cs

```diff
@@ -0,0 +1,1361 @@
+        var result = Call(TokenAddress, 0, method, new object[] { Message.Sender, saleInfo.TokenAmount });
+
+        Assert(result.Success && (bool)result.ReturnValue, "Token transfer failed.");
+
+        Log(new InvestLog { Sender = Message.Sender, Invested = saleInfo.Invested, TokenAmount = saleInfo.TokenAmount, Refunded = saleInfo.RefundAmount });
+
+        TokenBalance = checked(TokenBalance - saleInfo.TokenAmount);
+
+        if (saleInfo.RefundAmount > 0) // refund over sold amount
+            Transfer(Message.Sender, saleInfo.RefundAmount);
+
+        return true;
+    }
+
+    private void EnsureKycVerified()
+    {
+        var result = Call(MapperAddress, 0, "GetSecondaryAddress", new object[] { Message.Sender });
+
+        if (result.Success && result.ReturnValue is Address identityAddress && identityAddress != Address.Zero)
+        {
+            result = Call(KYCAddress, 0, "GetClaim", new object[] { identityAddress, (uint)3 /*shufti kyc*/ });
+
+            Assert(result.Success && result.ReturnValue != null, "Your KYC is not verified.");
+
+            return;
+        }
+
+        Assert(false, "The address has no mapping.");
+    }
+
+    public bool WithdrawFunds()
+    {
+        Assert(Message.Sender == Owner, "Only contract owner can transfer funds.");
+        Assert(!SaleOpen, "STO is not ended yet.");
+
+        var result = Transfer(this.Owner, Balance);
+
+        return result.Success;
+    }
+
+    public bool WithdrawTokens()
+    {
+        Assert(!IsNonFungibleToken, $"The {nameof(WithdrawTokens)} method is not supported for Non-Fungible Token.");
+        Assert(Message.Sender == Owner, "Only contract owner can transfer tokens.");
+        Assert(!SaleOpen, "STO is not ended yet.");
+
+        var result = Call(TokenAddress, 0, nameof(StandardToken.TransferTo), new object[] { Message.Sender, TokenBalance });
+
+        TokenBalance = 0;
+
+        Assert(result.Success && (bool)result.ReturnValue, "Token transfer failed.");
+
+        return true;
+    }
+
+    private SalePeriod GetCurrentPeriod()
+    {
+        var result = default(SalePeriod);
+
+        foreach (var period in SalePeriods)
+        {
+            if (period.EndBlock >= Block.Number)
+            {
+                result = period;
+                break;
+            }
+        }
+
+        return result;
+    }
+    private void SetPeriods(SalePeriodInput[] periods)
+    {
+        var salePeriods = ConvertSalePeriodInputs(periods);
+
+        SalePeriods = salePeriods;
+        EndBlock = salePeriods[salePeriods.Length - 1].EndBlock;
+    }
+
+    private SalePeriod[] ConvertSalePeriodInputs(SalePeriodInput[] periods)
+    {
+        var result = new SalePeriod[periods.Length];
+        var blockNumber = Block.Number;
+        for (int i = 0; i < periods.Length; i++)
+        {
+            var input = periods[i];
+            blockNumber = checked(blockNumber + input.DurationBlocks);
```

Comment:
Might be worth adding a check here to always make sure that 
```
if (i > 0) 
{
    Assert(input.DurationBlocks > periods[i - 1].DurationBlocks
}
```

Specifically in reference to `GetCurrentPeriod()` returning the first SalePeriod that is greater than `Block.Number` important to make sure these are in the correct order. Especially if 3rd parties deploy their own contracts and miss this validation client side on their end.

Reply:
These are durations and not block numbers so ordering is not needed. Each period specify duration as number of blocks and we calculate block numbers without breaking order and save it.